x/rewards/ante/fee_deduction.go

```diff
@@ -0,0 +1,129 @@
+func NewDeductFeeDecorator(ak ante.AccountKeeper, bk authTypes.BankKeeper, fk ante.FeegrantKeeper, rk RewardsKeeperExpected) DeductFeeDecorator {
+	return DeductFeeDecorator{
+		ak:             ak,
+		bankKeeper:     bk,
+		feegrantKeeper: fk,
+		rewardsKeeper:  rk,
+	}
+}
+
+// AnteHandle implements the ante.AnteHandler interface.
+func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkErrors.Wrap(sdkErrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	if addr := dfd.ak.GetModuleAddress(authTypes.FeeCollectorName); addr == nil {
+		return ctx, fmt.Errorf("fee collector module account (%s) has not been set", authTypes.FeeCollectorName)
+	}
+
+	fee := feeTx.GetFee()
+	feePayer := feeTx.FeePayer()
+	feeGranter := feeTx.FeeGranter()
+
+	deductFeesFrom := feePayer
+
+	// If feegranter set, deduct fee from feegranter account (only when feegrant is enabled)
+	if feeGranter != nil {
+		if dfd.feegrantKeeper == nil {
+			return ctx, sdkErrors.Wrap(sdkErrors.ErrInvalidRequest, "fee grants are not enabled")
+		}
+
+		if !feeGranter.Equals(feePayer) {
+			if err := dfd.feegrantKeeper.UseGrantedFees(ctx, feeGranter, feePayer, fee, tx.GetMsgs()); err != nil {
+				return ctx, sdkErrors.Wrapf(err, "%s not allowed to pay fees from %s", feeGranter, feePayer)
+			}
+		}
+
+		deductFeesFrom = feeGranter
+	}
+
+	deductFeesFromAcc := dfd.ak.GetAccount(ctx, deductFeesFrom)
+	if deductFeesFromAcc == nil {
+		return ctx, sdkErrors.Wrapf(sdkErrors.ErrUnknownAddress, "fee payer address (%s) does not exist", deductFeesFrom)
+	}
+
+	// Deduct the fees
+	if !feeTx.GetFee().IsZero() {
+		if err := dfd.deductFees(ctx, deductFeesFromAcc, feeTx.GetFee()); err != nil {
+			return ctx, err
+		}
+	}
+
+	events := sdk.Events{sdk.NewEvent(sdk.EventTypeTx,
+		sdk.NewAttribute(sdk.AttributeKeyFee, feeTx.GetFee().String()),
+	)}
+	ctx.EventManager().EmitEvents(events)
+
+	return next(ctx, tx, simulate)
+}
+
+// deductFees deducts fees from the given account if rewards calculation and distribution is enabled.
+// If rewards module is disabled, all the fees are sent to the fee collector account.
+// NOTE: this is the only logic being changed.
+func (dfd DeductFeeDecorator) deductFees(ctx sdk.Context, acc authTypes.AccountI, fees sdk.Coins) error {
+	// TODO: we need to identify Msg type to only deduct fees for the WASM operation (not all of them)
+
+	if !fees.IsValid() {
+		return sdkErrors.Wrapf(sdkErrors.ErrInsufficientFee, "invalid fee amount: %s", fees)
+	}
+
+	// Send everything to the fee collector account if rewards are disabled
+	if !dfd.rewardsKeeper.TxFeeRewardsEnabled(ctx) {
+		if err := dfd.bankKeeper.SendCoinsFromAccountToModule(ctx, acc.GetAddress(), authTypes.FeeCollectorName, fees); err != nil {
+			return sdkErrors.Wrapf(sdkErrors.ErrInsufficientFunds, err.Error())
+		}
+		return nil
```

Comment:
We need to be conscious about this approach for tracking pruning, because tracking can have records of this TX whilst rewards does not.

Reply:
Also side note, we're fetching the `Dec` twice here, for performance we could just fetch the tx fee rebate ratio here, and apply the `isZero()` logic here.

Reply:
`Dec` usage fixed. As for pruning. If there were no TxRebate rewards a corresponding rewards tracking entry is not created and this is handled by the `x/rewards` distribution logic.